sdk/tests/core/test_az_mqtt5_connection.c

```diff
@@ -434,28 +668,917 @@ static void test_az_mqtt5_connection_open_idle_success(void** state)
+
+  assert_int_equal(ref_conn_rsp_ok, 2);
+  assert_int_equal(ref_conn_error, 0);
+  assert_int_equal(ref_disconn_req, 2);
+}
+
+// Testing that when disconnecting, a connect response error from the broker leads to a faulted
+// state.
+static void test_az_mqtt5_connection_disconnecting_connect_rsp_err_failure(void** state)
+{
+  (void)state;
+  _reset_refs();
+  az_mqtt5_connection test_connection;
+  az_mqtt5 mock_mqtt;
+  az_mqtt5_options mock_mqtt_options = { 0 };
+  _az_event_client mock_client_1;
+  _az_event_client mock_client_2;
+  _az_hfsm mock_client_hfsm_1;
+  _az_hfsm mock_client_hfsm_2;
+
+  _az_mqtt5_connection_setup(
+      &test_connection,
+      &test_connection_options,
+      &mock_mqtt,
+      &mock_mqtt_options,
+      &mock_client_1,
+      &mock_client_2,
+      &mock_client_hfsm_1,
+      &mock_client_hfsm_2);
+
+  _az_mqtt5_connection_idle_connecting(&test_connection);
+  _az_mqtt5_connection_connecting_connected(&mock_mqtt);
+  _az_mqtt5_connection_connected_disconnecting(&test_connection);
+
+  assert_int_equal(
+      az_mqtt5_inbound_connack(
+          &mock_mqtt,
+          &(az_mqtt5_connack_data){ .connack_reason = AZ_MQTT5_CONNACK_UNSPECIFIED_ERROR,
+                                    .tls_authentication_error = false }),
+      AZ_OK);
+
+  assert_int_equal(ref_conn_rsp_ok, 2);
+  assert_int_equal(ref_conn_rsp_err, 0);
+  assert_int_equal(ref_conn_error, 1);
+  assert_int_equal(ref_disconn_req, 1);
+
+  _az_mqtt5_connection_faulted_check(&test_connection);
+}
+
+// Testing that when disconnecting, an ok connect response from the broker will lead to a second
+// disconnect request to be sent.
+static void test_az_mqtt5_connection_disconnecting_connect_rsp_ok_success(void** state)
+{
+  (void)state;
+  _reset_refs();
+  az_mqtt5_connection test_connection;
+  az_mqtt5 mock_mqtt;
+  az_mqtt5_options mock_mqtt_options = { 0 };
+  _az_event_client mock_client_1;
+  _az_event_client mock_client_2;
+  _az_hfsm mock_client_hfsm_1;
+  _az_hfsm mock_client_hfsm_2;
+
+  _az_mqtt5_connection_setup(
+      &test_connection,
+      &test_connection_options,
+      &mock_mqtt,
+      &mock_mqtt_options,
+      &mock_client_1,
+      &mock_client_2,
+      &mock_client_hfsm_1,
+      &mock_client_hfsm_2);
+
+  _az_mqtt5_connection_idle_connecting(&test_connection);
+  _az_mqtt5_connection_connecting_connected(&mock_mqtt);
+  _az_mqtt5_connection_connected_disconnecting(&test_connection);
+
+  will_return(__wrap_az_platform_clock_msec, 0);
+  will_return(__wrap_az_mqtt5_outbound_disconnect, AZ_OK);
+  assert_int_equal(
+      az_mqtt5_inbound_connack(
+          &mock_mqtt,
+          &(az_mqtt5_connack_data){ .connack_reason = AZ_MQTT5_CONNACK_SUCCESS,
+                                    .tls_authentication_error = false }),
+      AZ_OK);
+
+  assert_int_equal(ref_conn_rsp_ok, 2);
+  assert_int_equal(ref_conn_rsp_err, 0);
+  assert_int_equal(ref_conn_error, 0);
+  assert_int_equal(ref_disconn_req, 1);
```

Comment:
Should this be 2 if an additional disconnect request is sent?

Reply:
The `ref_disconn_req` variable only gets increased when the application calls connection close which isn't happening in this case

Reply:
Should it be renamed ref_conn_close_req then?